### PR TITLE
feat: dapp explorer

### PR DESCRIPTION
**Pull request by southbears**
## Description

Adds page for exploring dApps

### Questions
1. The WC icon is currently used in the wallet connection modal (where we want color) and also in this exploration page (where we don't want color). How to best handle scenarios where we want brand color vs not? Now I created two different icons, but that doesn't feel right
2. How to feature flag plugins? I don't want the `WalletConnectToDapps` plugin to be used if the `WalletConnectToDapps` feature flag is set. Atm my plugin is in the `activePlugins` list
3. Is there some place in the product where this kind of pagination component is used? I didn't add pagination yet as I wanted to check if there was already a component for it first 
image
4. Where do `Get Started` and `Learn More` CTAs in the banner link to?
5. The banner should prob have the image on top on mobile web. How would you do that? `<Grid>` with different `gridTemplateColumns` based on breakpoint?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)
https://app.dework.xyz/southbear/shapeshiftwalletcon/view/board-l7nqagyc?taskId=f0333b1a-9719-4d30-be97-99db38d23e16

### Engineering

Go to `Explore > dApps`

### Operations

Go to `Explore > dApps`

## Screenshots (if applicable)
image

### CODE REVIEW

File: src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const fs = require('fs')
+const path = require('path')
+const fetch = require('node-fetch')
+
+const alphabet = 'abcdefghijklmnopqrstuvwxyz'.split('')
+
+const registryJsonUrl =
+  'https://explorer.walletconnect.com/_next/data/p88ImN-VjsDejK2p4EYtd/registry.json'
+
+async function run() {
```
Comment (southbears) [inline src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js:10]: Script I ran locally to generate `registry.json`

**Comment by DiggyDiggy2** [comment]
@southbears in regard to question #3 - this is a new component, thats not currently used elsewhere yet

**Comment by DiggyDiggy2** [comment]
@southbears in regards to question #4. 
The "get started" button opens this modal: https://www.figma.com/proto/OHbVrAVaV2xQWUsq0nKzGm/ShapeShift-Library?node-id=1572%3A6756&scaling=min-zoom&page-id=1222%3A4033&starting-point-node-id=1296%3A5004
The "learn more" will likely open an article on ZenDesk which will go into more depth about the feature (not created yet)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 <!-- Before submitting a pull request, please make sure you have answered the following: -->
 
-- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
+- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
```
Comment (0xdef1cafe) [inline .github/PULL_REQUEST_TEMPLATE.md:9]: thanks for fixing this - do you mind doing this in a separate PR though?

File: src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const fs = require('fs')
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js:1]: can we please move this to the `/scripts` folder such that it's not included into the build? also we may want to convert to typescript and invoke it via a yarn script with `ts-node` - see how we do `yarn release` for reference

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+              type='text'
+              placeholder='Search'
+              pl={10}
+              variant='filled'
+            />
+          </InputGroup>
+        </Box>
+      </Stack>
+      <SimpleGrid columns={{ lg: 4, sm: 2, base: 1 }} spacing={4}>
+        {filteredListings.slice(page * pageSize, pageSize).map(listing => (
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:70]: `slice` creates a new ref every single time - can we please rip this out and `useMemo` this part of the tree?

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
hey @southbears this is actually looking pretty good, can you please

a) rebase/merge develop, and b) address some small requested changes and i'll rereview

**Comment by 0xdef1cafe** [comment]
w.r.t 5. see here how we set `display` css props on different breakpoints - you can do this with any of the Chakra UI components

https://github.com/shapeshift/web/blob/develop/src/pages/Dashboard/Portfolio.tsx#L126

**Comment by 0xdef1cafe** [comment]
for 2. https://github.com/shapeshift/web/blob/develop/src/plugins/avalanche/index.tsx#L15 this is how we feature flag a plugin

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (gomesalexandre) [reply .github/PULL_REQUEST_TEMPLATE.md]: This has already been added in https://github.com/shapeshift/web/pull/2650/files#diff-18813c86948efc57e661623d7ba48ff94325c9b5421ec9177f724922dd553a35R9

@southbears just merged develop from GitHub UI - this will make it easier to reviewers to get this one to the finish line

File: src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js (hunk 3 above)
Comment (gomesalexandre) [reply src/plugins/walletConnectToDapps/scripts/scrapeWalletConnectRegistry.js:1]: Additionally, and unless using a package which doesn't support these (which shouldn't be the case, since this is `require`ing the Node.JS standard library and node-fetch both of which should be compatible), please use [ES Modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) and not CommonJS ones - as we conventionally use ESM around the web codebase.

File: src/components/Icons/WalletConnectIcon.tsx [hunk 5]
```diff
@@ -17,3 +17,16 @@ export const WalletConnectIcon = createIcon({
   ),
   viewBox: '0 0 300 185',
 })
+
+export const WalletConnectCurrentColorIcon = createIcon({
+  displayName: 'WalletConnectIcon',
```
Comment (gomesalexandre) [inline src/components/Icons/WalletConnectIcon.tsx:22]: This has the same displayName as the existing `WalletConnectIcon`, let's keep it consistent with the const name

File: src/pages/Dashboard/RecentTransactions.tsx [hunk 6]
```diff
@@ -1,7 +1,7 @@
-import { useSelector } from 'react-redux'
 import { Card } from 'components/Card/Card'
 import { Text } from 'components/Text'
 import { TransactionHistoryList } from 'components/TransactionHistory/TransactionHistoryList'
+import { useSelector } from 'react-redux'
```
Comment (gomesalexandre) [inline src/pages/Dashboard/RecentTransactions.tsx:4]: Revert this hunk, if you run `yarn lint --fix` you can see this doesn't pass lint and will get autofixed back to its original line

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  Link,
+  SimpleGrid,
+  Stack,
+  Text as PlainText
+} from '@chakra-ui/react'
+import { Text } from 'components/Text'
+import { FC, useEffect, useMemo } from 'react'
+import { useForm } from 'react-hook-form'
+
+interface RegistryItem {
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:18]: Not a blocker but we conventionally use types vs. interfaces for most use-cases

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  category: string
+  id: string
+  homepage: string
+  name: string
+  image: string
+}
+
+const registry: RegistryItem[] = require('../registry.json')
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:26]: We can import JSON files with the ESM `import` syntax as well

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  const { register, watch, setValue } = useForm<{ search: string; page: number }>({
+    mode: 'onChange',
+    defaultValues: {
+      search: '',
+      page: 0,
+    },
+  })
+
+  const search = watch('search')
+  const page = watch('page')
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:40]: Use `useWatch()` instead of watch: the former is react-friendly, the latter isn't and will require you to unsubscribe to watched inputs
https://react-hook-form.com/api/useform/watch

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 8 above)
Comment (gomesalexandre) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:26]: To honor the type name and make it easier to name items iterated over

```suggestion
const registryItems: RegistryItem[] = require('../registry.json')
```

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  useEffect(() => setValue('page', 0), [search, setValue])
+
+  const filteredListings = useMemo(
+    () => registry.filter(l => !search || l.name.toLowerCase().includes(search.toLowerCase())),
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:44]: Use full variable names in callbacks 🙏🏽 

```suggestion
    () => registryItems.filter(registryItem => !search || registryItem.name.toLowerCase().includes(search.toLowerCase())),
```

File: src/plugins/walletConnectToDapps/components/ExplorationBanner.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+        <Stack direction='row'>
+          <Button colorScheme='blue' size='sm' onClick={() => alert('get started :)')}>
+            <Text translation='plugins.walletConnectToDapps.getStarted' />
+          </Button>
+          <Button
+            colorScheme='blue'
+            size='sm'
+            variant='ghost'
+            onClick={() => alert('learn more :)')}
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/ExplorationBanner.tsx:36]: If the `onClick` handler is going to be implemented in a follow-up PR, add a TODO for it

File: src/plugins/walletConnectToDapps/components/ExplorationBanner.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+          icon={<WalletConnectCurrentColorIcon color={useColorModeValue('blue.500', 'blue.400')} />}
+        />
+
+        <Box>
+          <Text as='b' fontSize='lg' translation='plugins.walletConnectToDapps.banner.title' />
+          <Text translation='plugins.walletConnectToDapps.banner.subtitle' color='gray.500' />
+        </Box>
+
+        <Stack direction='row'>
+          <Button colorScheme='blue' size='sm' onClick={() => alert('get started :)')}>
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/ExplorationBanner.tsx:29]: If the `onClick` handler is going to be implemented in a follow-up PR, add a TODO for it

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            <InputLeftElement pointerEvents='none'>
+              <SearchIcon color='gray.700' />
+            </InputLeftElement>
+            <Input
+              {...register('search')}
+              type='text'
+              placeholder='Search'
+              pl={10}
+              variant='filled'
+            />
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:65]: @shapeshift/product this has autocomplete from the asset search input, do we want to turn autocomplete off for it?

image

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+          </InputGroup>
+        </Box>
+      </Stack>
+      <SimpleGrid columns={{ lg: 4, sm: 2, base: 1 }} spacing={4}>
+        {filteredListings.slice(page * pageSize, pageSize).map(listing => (
+          <Link key={listing.id} href={listing.homepage} isExternal>
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:71]: @shapeshift/product are we happy with this hover style?

image

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+  category: string
+  id: string
+  homepage: string
+  name: string
+  image: string
+}
+
+const registry: RegistryItem[] = require('../registry.json')
+
+const pageSize = 20
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:28]:
```suggestion
const PAGE_SIZE = 20
```

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
This looks good so far @southbears, well done!

In addition to the great review from @0xdef1cafe, added some additional comments WRT our conventions, optimizations, as well as some questions to be answered by product.

**Comment by gomesalexandre** [comment]
Also, could you please set-up eslint in your editor of choice or run `yarn lint:fix`? Your diff has a few lint violations that are autofixable.

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 8 above)
Comment (southbears) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx]: @gomesalexandre how would you get the variable typed properly when using import?

**Comment by southbears** [comment]
@gomesalexandre the errors here aren't fixable with `yarn lint --fix`: https://github.com/shapeshift/web/runs/8294975995?check_suite_focus=true
Any guidance?

File: src/plugins/walletConnectToDapps/components/PageInput.tsx [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+      borderRadius='md'
+      borderColor={useColorModeValue('blackAlpha.50', 'gray.750')}
+    >
+      <IconButton
+        colorScheme='gray'
+        aria-label='Previous Page'
+        variant='ghost'
+        icon={<ArrowBackIcon />}
+        disabled={value <= 0}
+        style={{ height: 38 }}
```
Comment (southbears) [inline src/plugins/walletConnectToDapps/components/PageInput.tsx:28]: @gomesalexandre how would you think about this? I want this to have the same height as the search bar input (40px). However, since this is wrapped in a HStack with a border, the UI wouldn't be properly aligned. Height 38 makes it align properly

image

Comment (gomesalexandre) [reply src/plugins/walletConnectToDapps/components/PageInput.tsx:28]: Fixed pxs are something to be cautious about but given this looks good both on desktop and mobile, I don't see any issue with it. Summoning @reallybeard for CSS wizardry answer

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 14 above)
Comment (reallybeard) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:69]: If possible it would be good to increase the opacity of the blurred background image on hover from 0.3 to 0.5. I'm not against the underline for now.

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 13 above)
Comment (reallybeard) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:62]: yah turn off auto complete for this input please.

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 4 above)
Comment (southbears) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx]: Does that matter? This is just mapping over a certain set of elements (`listing`) which are objects that don't change, regardless of if they're sliced. Aka `const array = [{}]; array[0] === array.slice()[0] ===> true`. More than happy to memoize though, I'm just not sure it matters performance-wise

**Comment by 0xdef1cafe** [comment]
> @gomesalexandre the errors here aren't fixable with `yarn lint --fix`: https://github.com/shapeshift/web/runs/8294975995?check_suite_focus=true Any guidance?

it's best to configure your IDE to use eslint and our `.eslintrc` and it will "just work"

it's easy to just delete the imports then use usual IDE tools to find and reimport them

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 13 above)
Comment (southbears) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:62]: Disabled autocomplete

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 14 above)
Comment (southbears) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:69]: Done

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+              placeholder='Search'
+              pl={10}
+              variant='filled'
+            />
+          </InputGroup>
+        </Box>
+        <PageInput value={page} max={maxPage} onChange={value => setValue('page', value)} />
+      </Stack>
+      <SimpleGrid columns={{ lg: 4, sm: 2, base: 1 }} spacing={4}>
+        {filteredListings.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE).map(listing => (
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:70]: @reallybeard we want some sort of empty state here too

image

File: src/plugins/walletConnectToDapps/components/PageInput.tsx [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+        colorScheme='gray'
+        aria-label='Previous Page'
+        variant='ghost'
+        icon={<ArrowBackIcon />}
+        disabled={value <= 0}
+        style={{ height: 36 }}
+        onClick={prevPage}
+      />
+      <Text textAlign='center' minWidth={16}>
+        {value + 1} of {max + 1}
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/PageInput.tsx:33]: translations

File: src/plugins/walletConnectToDapps/components/PageInput.tsx [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+  return (
+    <HStack
+      align='center'
+      borderWidth={2}
+      borderRadius='md'
+      borderColor={useColorModeValue('blackAlpha.50', 'gray.750')}
+    >
+      <IconButton
+        colorScheme='gray'
+        aria-label='Previous Page'
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/PageInput.tsx:25]: translations

File: src/plugins/walletConnectToDapps/components/PageInput.tsx [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+        icon={<ArrowBackIcon />}
+        disabled={value <= 0}
+        style={{ height: 36 }}
+        onClick={prevPage}
+      />
+      <Text textAlign='center' minWidth={16}>
+        {value + 1} of {max + 1}
+      </Text>
+      <IconButton
+        aria-label='Next Page'
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/PageInput.tsx:36]: translations

File: src/plugins/walletConnectToDapps/components/PageInput.tsx [hunk 21]
```diff
@@ -0,0 +1,45 @@
+import { ArrowBackIcon, ArrowForwardIcon } from '@chakra-ui/icons'
+import { HStack, IconButton, Text, useColorModeValue } from '@chakra-ui/react'
+import type { FC } from 'react'
+import { useCallback } from 'react'
+
+interface Props {
```
Comment (0xdef1cafe) [inline src/plugins/walletConnectToDapps/components/PageInput.tsx:6]: type over interface for consistency please

File: package.json [hunk 22]
```diff
@@ -28,6 +28,7 @@
     "build-storybook": "build-storybook -s public",
     "postinstall": "./scripts/postinstall.sh",
     "release": "ts-node --skipProject ./scripts/release.ts",
+    "scrape-walletconnect-dapps": "ts-node --skipProject ./scripts/scrapeWalletConnectRegistry.ts",
```
Comment (0xdef1cafe) [inline package.json:31]: thank you!

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
very small changes at this point, looking good

File: scripts/scrapeWalletConnectRegistry.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+
+const alphabet = 'abcdefghijklmnopqrstuvwxyz'.split('')
+
+const registryJsonUrl =
+  'https://explorer.walletconnect.com/_next/data/CWVJGGRcXLbAilG1CUAn3/index.json'
+
+async function run() {
+  const listings: RegistryItem[] = []
+  for (const searchQuery of ['', ...alphabet]) {
+    // eslint-disable-next-line no-console
```
Comment (gomesalexandre) [inline scripts/scrapeWalletConnectRegistry.ts:14]: Note for other reviewers: This is actually correct, we don't want to use `shapeshiftoss/logger/no-native-console` as it will hint about moduleLogger use, which we can't use here

**Review by gomesalexandre — COMMENTED** [review]
LGTM, retested and still looking good, autocomplete is off, and the scrape script is abstracted to a `.ts` file with a matching package.json `scripts` entry.

Awaiting for the empty state of the `filteredListings.slice().map()` expression and we're ready to merge this 👏🏽 

Also @southbears, can you update the engineering testing steps to reflect the addition of the `scrape-walletconnect-dapps` script? 🙏🏽

File: src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx (hunk 17 above)
Comment (reallybeard) [reply src/plugins/walletConnectToDapps/components/DappRegistryGrid.tsx:70]: Screen Shot 2022-09-20 at 4 11 26 PM

**Review by 0xdef1cafe — APPROVED** [review]
